### ISSUE TITLE
Mocha runner does not support require "esm"

### ISSUE THREAD

**Issue by nicojs** [stryker-mutator/stryker-js#3014]
**Summary**

When using a mocha project with esm, you might run into this error:

```
20:15:50 (21474) ERROR DryRunExecutor One or more tests resulted in an error:
        SyntaxError: Cannot use import statement outside a module
/home/nicojs/github/mocha-esm-ts-node/.stryker-tmp/sandbox7723690/test/say-hello.spec.ts:2
import assert from 'assert';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at new Script (vm.js:102:7)
    at createScript (vm.js:263:10)
    at Object.runInThisContext (vm.js:311:10)
    at wrapSafe (internal/modules/cjs/loader.js:989:15)
    at Module._compile (internal/modules/cjs/loader.js:1049:27)
```

The reason for this is that mocha has an exception in place for `--require esm`:

https://github.com/mochajs/mocha/blob/6eb3c3cda72102144ef2f2dd9b90ddc2d074c9b6/bin/mocha#L88-L94

In order to make this work in Stryker, you would need to rebuild this yourself

**Stryker config**

```json
{
  "$schema": "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
  "testRunner": "mocha",
  "mochaOptions": {
    "spec": ["test/*.ts", "test/*.js"],
    "require": ["esm", "ts-node/register"]
  }
}

```

**Test runner config**

```
env TS_NODE_PROJECT=\"tsconfig.json\" mocha -r ts-node/register -r esm \"test/*.{js,ts}\"
```

**Stryker environment**

```
+-- @stryker-mutator/core@5.2.2
+-- @stryker-mutator/mocha-runner@5.2.2
```

(5.2.2 or earlier)

**Comment by nicojs** [comment]
The solution is to change your Stryker config to also require `esm` the same way as mocha does it:

```diff
{
  "$schema": "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
  "testRunner": "mocha",
+ "testRunnerNodeArgs": ["--require", "esm"],

  "mochaOptions": {
    "spec": ["test/*.ts", "test/*.js"],
-   "require": ["esm", "ts-node/register"]
+   "require": ["ts-node/register"]
  }
}
```

We should throw an error from the mocha runner, explaining how to fix it.

**Comment by nicojs** [comment]
Maybe interesting to know: `esm` support is dropped from mocha as of v9. See

https://github.com/mochajs/mocha/issues/4671

**Comment by nicojs** [comment]
We released `v5.2.3` with the fix
